docs: add ma v2 page (#1340)
alchemyplatform/aa-sdk: site/pages/smart-contracts/modular-account-v2/getting-started.mdx

---
title: Modular Account V2 • Getting started
description: Getting started with Modular Account V2 in Account Kit
---

# Getting started with Modular Account V2

> 

Author: there'll be an overview page before this, i was thinking that that'll be the page that includes the benefits. does that layout make sense?

> 

Author: overview page added here: https://github.com/alchemyplatform/aa-sdk/pull/1345

> It is easy to get started with Modular Account V2! We will show you two different ways using Alchemy Infra or 3rd party infra.

Reviewer: ```suggestion
It is easy to get started with Modular Account v2! Below, you will create a new Modular Account v2 client that will be used to send user operations. Your MAv2 smart account will be deployed on-chain when you send the first User Operation from a unique signer.
```

## Install packages

**Prerequisites**

- minimum Typescript version of 5
- pin viem to 2.20.0 (`yarn add viem@2.20.0`)

**Installation**

First, install the `@account-kit/smart-contracts` package.

:::code-group

```bash [yarn]
yarn add @account-kit/smart-contracts

> # if using alchemy infra

Reviewer: ```suggestion
```

yarn add @account-kit/infra
```

```bash [npm]
yarn add @account-kit/smart-contracts
# if using alchemy infra
yarn add @account-kit/infra

> ```

Reviewer 2: should be `npm` instead of `yarn` in this example?

Author: good catch. this is from ma v1 lol, we need to change both

:::

:::tip[Address calculation]
For Modular Account V2, the address of the smart account will be calculated as a combination of [the owner and the salt](https://github.com/alchemyplatform/modular-account/blob/v2.0.x/src/factory/AccountFactory.sol#L98-L104). You will get the same smart account address each time you supply the same `owner`, the signer(s) used to create the account for the first time. You can also optionally supply `salt` if you want a different address for the same `owner` param (the default salt is `0n`).

If you want to use a signer to connect to an account whose address does not map to the contract-generated address, you can supply the `accountAddress` to connect with the account of interest. In that case, the `signer` address is not used for address calculation, but only for signing the operation.
:::

## Creating a Modular Account V2 client

```ts twoslash [modular-account-v2.ts]

> import { createModularAccountV2Client } from "@account-kit/smart-contracts";

Author: yeah it should be, was merged into main from adam's 7702-core branch

> import { createModularAccountV2Client } from "@account-kit/smart-contracts";

Author: `createSMAV2AccountClient` is old, u might need to pull from main

import { LocalAccountSigner } from "@aa-sdk/core";
import { sepolia, alchemy } from "@account-kit/infra";
import { generatePrivateKey } from "viem/accounts";

> const accountClient = await createModularAccountV2Client({

Reviewer: I still feel like we should show the modes param here

> const accountClient = await createModularAccountV2Client({

Reviewer: Maybe we have a comment after this snippet just mentioning that the "mode" param is an optional param that can be used to specify the MAv2 variant you'd like to use. Supported options are 
- "default": uses the base version of MAv2 and will default to this if no mode passed
- "7702": read more about using EIP-7702 compatible accounts to upgrade EOAs (link to adam guide)

  chain: sepolia,

>   transport: alchemy({ apiKey: "your-api-key" }), // or http("RPC_URL") for non alchemy infra

Reviewer: ```suggestion
  transport: alchemy({ apiKey: "your-api-key" }), // Get your API key at https://dashboard.alchemy.com/apps or http("RPC_URL") for non-alchemy infra
```

  signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey()),
});
```

> Next, if you want to use a different `signer` with a smart account signer, check out [choosing a signer](/signer/what-is-a-signer). Otherwise, if you are ready to get on-chain, go to [send user operations](/infra/send-user-operations).

Reviewer: ```suggestion
Want to enable social login methods? Set up your [Alchemy Signer](/signer/quickstart).  

Alternatively, you can [bring a 3rd party signer](/third-party/signers) as the owner of your new account. 

Not sure what signer to use? [Learn more](/signer/what-is-a-signer).
```

## Sending a user operation

> Now that you have a client, you can send a user operation with Modular Account V2.

Reviewer: ```suggestion
Now that you have a client, you can send a User Operation. The first User Operation will also deploy the new Modular Account v2. 
```

```ts twoslash
import { createModularAccountV2Client } from "@account-kit/smart-contracts";
import { LocalAccountSigner } from "@aa-sdk/core";
import { sepolia, alchemy } from "@account-kit/infra";
import { generatePrivateKey } from "viem/accounts";

const accountClient = await createModularAccountV2Client({
  chain: sepolia,
  transport: alchemy({ apiKey: "your-api-key" }),
  signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey()),
});

const operation = await accountClient.sendUserOperation({
  target: "0x123..",
  data: "0x123..",
  value: "0n",

> });

Reviewer: ```suggestion
});

console.log(
"User operation sent! \nUO hash: ",
operation.hash,
"\nModular Account v2 Address: ",
operation.request.sender
);
```

For Modular Account V2, the address of the smart account will be calculated as a combination of [the owner and the salt](https://github.com/alchemyplatform/modular-account/blob/v2.0.x/src/factory/AccountFactory.sol#L98-L104). You will get the same smart account address each time you supply the same `owner`, the signer(s) used to create the account for the first time. You can also optionally supply `salt` if you want a different address for the same `owner` param (the default salt is `0n`).

If you want to use a signer to connect to an account whose address does not map to the contract-generated address, you can supply the `accountAddress` to connect with the account of interest. In that case, the `signer` address is not used for address calculation, but only for signing the operation.
:::

## Creating a Modular Account V2 client

```ts twoslash [modular-account-v2.ts]
import { createModularAccountV2Client } from "@account-kit/smart-contracts";
import { LocalAccountSigner } from "@aa-sdk/core";
import { sepolia, alchemy } from "@account-kit/infra";
import { generatePrivateKey } from "viem/accounts";

const accountClient = await createModularAccountV2Client({

>   chain: sepolia,

Reviewer: ```suggestion
  mode: "default", // optional param to specify the MAv2 variant (either "default" or "7702")
  chain: sepolia,
```
